File: carp.dsp.core/src/jvmTest/resources/scripts/process_data.R
#!/usr/bin/env Rscript
# Simple deterministic data processing script
# Reads input CSV, adds a computed column, writes output CSV

# Get command-line arguments
args <- commandArgs(trailingOnly = TRUE)

if (length(args) < 2) {
  stop("Usage: process_data.R <input_file> <output_file>")
}

input_file <- args[1]
output_file <- args[2]

# Create output directory if needed
dir.create(dirname(output_file), showWarnings = FALSE, recursive = TRUE)

tryCatch({
  # Read input CSV
  if (file.exists(input_file)) {
    data <- read.csv(input_file, stringsAsFactors = FALSE)
    message(sprintf("Loaded %d rows from %s", nrow(data), input_file))
  } else {
    message(sprintf("Warning: Input file %s not found. Creating minimal output.", input_file))
    data <- data.frame(id = 1, value = 10)
  }
  
  # Process: add computed column
  if ("value" %in% names(data)) {
    data$processed_value <- as.numeric(data$value) * 2  # Simple: double the value
  } else {
    data$processed_value <- 0
  }
  
  # Write output CSV
  write.csv(data, output_file, row.names = FALSE)
  message(sprintf("Processed %d rows to %s", nrow(data), output_file))
  message(sprintf("Successfully processed %s -> %s", input_file, output_file))
  
}, error = function(e) {
  message(sprintf("Error processing data: %s", conditionMessage(e)))
  # Create minimal output on error
  data <- data.frame(id = 1, value = 10, processed_value = 20)
  write.csv(data, output_file, row.names = FALSE)
  message(sprintf("Created minimal output: %s", output_file))
})

File: carp.dsp.core/src/jvmTest/resources/scripts/process_data.py
#!/usr/bin/env python
"""
Simple deterministic data processing script.
Reads input CSV, adds a computed column, writes output CSV.
"""

import sys
import csv
from pathlib import Path

def process_data(input_file: str, output_file: str) -> None:
    """
    Process CSV file by reading and writing with a computed column.
    
    Args:
        input_file: Path to input CSV
        output_file: Path to output CSV
    """
    input_path = Path(input_file)
    output_path = Path(output_file)
    
    # Create parent directories if needed
    output_path.parent.mkdir(parents=True, exist_ok=True)
    print(f"Processing data from {input_file} to {output_file}...")
    
    # If input doesn't exist, create minimal output
    if not input_path.exists():
        print(f"Warning: Input file {input_file} not found. Creating minimal output.", file=sys.stderr)
        with open(output_path, 'w', newline='') as f:
            writer = csv.writer(f)
            writer.writerow(['id', 'value', 'processed_value'])
            writer.writerow([1, 10, 20])
        return
    
    # Read input and process
    try:
        with open(input_path, 'r', newline='') as infile:
            reader = csv.DictReader(infile)
            rows = list(reader)
        
        # Get fieldnames from input
        if not rows:
            fieldnames = ['id', 'value', 'processed_value']
            out_rows = []
        else:
            fieldnames = list(rows[0].keys()) + ['processed_value']
            out_rows = []
            for row in rows:
                # Add a computed column
                try:
                    value = float(row.get('value', 0))
                    row['processed_value'] = str(value * 2)  # Simple: double the value
                except (ValueError, TypeError):
                    row['processed_value'] = '0'
                out_rows.append(row)
        
        # Write output
        with open(output_path, 'w', newline='') as outfile:
            writer = csv.DictWriter(outfile, fieldnames=fieldnames)
            writer.writeheader()
            writer.writerows(out_rows)
        
        print(f"Processed {len(out_rows)} rows from {input_file} to {output_file}")
        
    except Exception as e:
        print(f"Error processing data: {e}", file=sys.stderr)
        # Create minimal output on error
        with open(output_path, 'w', newline='') as f:
            writer = csv.writer(f)
            writer.writerow(['id', 'value', 'processed_value'])
            writer.writerow([1, 10, 20])

if __name__ == '__main__':
    if len(sys.argv) < 3:
        print(f"Usage: {sys.argv[0]} <input_file> <output_file>")
        sys.exit(1)
    
    input_file = sys.argv[1]
    output_file = sys.argv[2]
    
    process_data(input_file, output_file)
    print(f"Successfully processed {input_file} -> {output_file}")
